hipThreads Linux nightly fails at CMake configure: libhipcxx declares an amdsmi runtime dep that doesn't exist when CORE_AMDSMI is off

## Summary

The hipThreads Linux nightly builds have been failing at CMake configure since Jul 29. `libhipcxx` declares `amdsmi` as an unconditional Linux runtime dependency, but the `amdsmi` target only exists when `THEROCK_ENABLE_CORE_AMDSMI` is on. The hipThreads CI group configures narrow with `THEROCK_ENABLE_ALL=OFF`, so the target never exists and configure dies before anything compiles.

Fix is up in #7034.

## Symptom

Both Linux hipThreads build jobs fail in the `Configure Projects` step. About fifteen instances of this, from various call sites in `therock_subproject.cmake`:

```
CMake Error at cmake/therock_subproject.cmake:1388 (get_target_property):
  get_target_property() called with non-existent target "amdsmi".
Call Stack (most recent call first):
  cmake/therock_subproject.cmake:1510 (_therock_assert_is_cmake_subproject)
  math-libs/CMakeLists.txt:524 (therock_cmake_subproject_declare)
```

Affected jobs in the `ROCm/rocm-libraries` TheRock CI Nightly:

- `Linux (hipthreads | gfx94X-dcgpu) / Build (gfx94X-dcgpu)`
- `Linux (hipthreads | gfx950-dcgpu) / Build (gfx950-dcgpu)`

Latest run: https://github.com/ROCm/rocm-libraries/actions/runs/30614819132

Two things in those jobs are cascade noise rather than separate problems. The `Report` step failing with `du: cannot access 'TheRock/build/dist/rocm'` and `Post Build Upload` failing with `FileNotFoundError: therock_manifest.json not found` are both just consequences of nothing having been built. The `Executing the custom container implementation failed. Please contact your self hosted runner administrator` annotations are the container wrapper's generic message on any non-zero exit, which makes this look like a runner or infrastructure problem when it isn't.

## Root cause

`math-libs/CMakeLists.txt` currently adds the dep based only on platform:

```cmake
set(libhipcxx_runtime_deps)
if(NOT WIN32)
  # Required on Linux.
  list(APPEND libhipcxx_runtime_deps amdsmi)
endif()
```

The `amdsmi` target is only declared under `THEROCK_ENABLE_CORE_AMDSMI` (see `core/CMakeLists.txt`). In a full build that's always on, so the dep resolves fine and this has never been a problem. In the hipThreads CI group it is off. From the failing configure log:

```
-- Implicitly enabled features: CORE_RUNTIME HIP_RUNTIME COMPILER LIBHIPCXX KPACK SYSDEPS BASE
--   * LIBHIPCXX (-DTHEROCK_ENABLE_LIBHIPCXX=ON)
--   * HIPTHREADS (-DTHEROCK_ENABLE_HIPTHREADS=ON)
-- Disabled features:
--   * CORE_AMDSMI (-DTHEROCK_ENABLE_CORE_AMDSMI=OFF)
```

`HIPTHREADS` implicitly pulls in `LIBHIPCXX`, but nothing pulls in `CORE_AMDSMI`, so `libhipcxx` ends up depending on a target that was never declared.

## Reproduce

```
cmake -B build -GNinja . \
  -DTHEROCK_AMDGPU_FAMILIES=gfx950-dcgpu \
  -DTHEROCK_ENABLE_HIPTHREADS=ON \
  -DTHEROCK_ENABLE_ALL=OFF
```

## Regression window

The unconditional dep landed in #6389 (merged Jul 27), which was correct for every configuration that existed at the time. It became reachable when the hipThreads build recipe landed in #6769 (Jul 28) and the nightly matrix entry landed in ROCm/rocm-libraries#9705 (Jul 29). The Jul 28 nightly was the last green one. It has been red on Jul 29, 30, and 31.

Worth being clear that no individual change here was wrong on its own. The dep was fine until a narrow configuration existed, and the narrow configuration couldn't have known about the dep.

## Precedent

This exact problem was already solved for the prim libs, in the same file, and the comment on it describes this scenario almost word for word:

```cmake
# amdsmi is Linux-only and only declared when THEROCK_ENABLE_CORE_AMDSMI is ON
# (see core/CMakeLists.txt). Benchmarks depend on primbench which requires
# amd_smi, so they must be disabled when amdsmi is unavailable (e.g., the
# prim-isolated CI group THEROCK_ENABLE_ALL=OFF / THEROCK_ENABLE_PRIM=ON).
if(WIN32 OR NOT TARGET amdsmi)
```

So the guard idiom exists and is documented, it just isn't enforced anywhere. My worry is that the next narrow CI group added will hit this again, since nothing stops a new recipe from copying the older platform-only pattern.

## Follow-ups

1. `hipthreads` isn't registered in `repos-config.json`, so TheRock CI skips PRs touching `projects/hipthreads` (ROCm/rocm-libraries#10218). As it stands the fix won't be exercised on the PR and we'll confirm at the next nightly.
2. It might be worth a general rule that any recipe depending on an optionally-enabled target guards on `TARGET`, or that the feature graph makes the implication explicit so the dep can stay unconditional. I don't have a strong opinion on which, and I'd defer to whoever owns the subproject machinery.
3. A PR that adds a CI matrix entry can't validate itself today. ROCm/rocm-libraries#9705 only touched `therock_matrix.py`, so nothing in it ran the new configuration.

--- a/math-libs/CMakeLists.txt
+++ b/math-libs/CMakeLists.txt
@@ -516,7 +516,7 @@ if(THEROCK_ENABLE_LIBHIPCXX)
   ##############################################################################
 
   set(libhipcxx_runtime_deps)
-  if(NOT WIN32)
+  if(NOT WIN32 AND TARGET amdsmi)
     # Required on Linux.
     list(APPEND libhipcxx_runtime_deps amdsmi)
   endif()

--- a/math-libs/CMakeLists.txt
+++ b/math-libs/CMakeLists.txt
@@ -516,8 +516,12 @@ if(THEROCK_ENABLE_LIBHIPCXX)
   ##############################################################################
 
   set(libhipcxx_runtime_deps)
+  # Used on Linux when amd-smi is available. amdsmi is only declared when
+  # THEROCK_ENABLE_CORE_AMDSMI is ON (see core/CMakeLists.txt), so guard on the
+  # target existing to keep narrow configs (e.g. THEROCK_ENABLE_LIBHIPCXX=ON /
+  # THEROCK_ENABLE_ALL=OFF, as in the hipThreads nightly) from referencing a
+  # non-existent target.
   if(NOT WIN32 AND TARGET amdsmi)
-    # Required on Linux.
     list(APPEND libhipcxx_runtime_deps amdsmi)
   endif()
 

